### PR TITLE
[DTRA] Maryia/DTRA-1559/feat: add isVerticalScrollEnabled prop to disable vertical scroll outside Y-axis

### PR DESCRIPTION
**Pull request by maryia-deriv**
To introduce the new `isVerticalScrollEnabled` prop which, when set to `false`, is working as follows:
- on Wheel devices - disables pointer events on the chart when a user is scrolling vertically outside Y-axis (this approach is not applicable to touch devices),
- on Touch devices - forces scroll on the nearest scrollable parent of the chart.

### CODE REVIEW

File: src/store/ChartAdapterStore.ts [hunk 1]
```diff
@@ -216,15 +248,87 @@ export default class ChartAdapterStore {
...
+            const { screenX, screenY } = e.touches[0];
+            this.touchValues = { x: screenX, y: screenY };
+        }
+    }
+    onTouchEnd(e: TouchEvent) {
+        const chartNode = this.mainStore.chart.chartNode;
+        if (chartNode && !this.mainStore.state.isVerticalScrollEnabled && e.touches.length === 1) {
+            this.touchValues = { yOnTouchEnd: e.touches[0].screenY };
+        }
+    }
```
Comment (maryia-deriv) [inline src/store/ChartAdapterStore.ts:263]: i can unite onTouchStart & onTouchEnd into one callback. what do you think?

File: src/store/ChartAdapterStore.ts [hunk 2]
```diff
@@ -216,15 +248,87 @@ export default class ChartAdapterStore {
...
+        if (e.type === 'wheel') e.preventDefault();
+
+        // Prevent vertical scroll on the chart if isVerticalScrollEnabled is false:
+        const chartNode = this.mainStore.chart.chartNode;
+        if (chartNode && !this.mainStore.state.isVerticalScrollEnabled) {
+            const nonScrollableAreaWidth = chartNode.offsetWidth - this.mainStore.chart.yAxisWidth;
+            const { left } = chartNode.getBoundingClientRect();
+
+            if (e.type === 'touchmove') {
+                // Touch devices approach - forcing scroll on a scrollable parent of the chart:
```
Comment (maryia-deriv) [inline src/store/ChartAdapterStore.ts:289]: i can move this logic for touch devices to its own callback (onTouchMove). what do you think?

File: src/store/ChartAdapterStore.ts [hunk 3]
```diff
@@ -439,24 +514,24 @@ export default class ChartAdapterStore {
...
                     if (price) delta_y = barNext.Close - price;
                 } else if (barPrev && barPrev.Close) {
-                    delta_x = x - this.getXFromEpoch(barPrev.DT!.getTime());
+                    delta_x = x - this.getXFromEpoch((barPrev.DT as Date).getTime());
 
                     ratio =
-                        (((date as unknown) as number) - bar.DT!.getTime()) /
-                        (bar.DT!.getTime() - barPrev.DT!.getTime());
+                        ((date as unknown as number) - (bar.DT as Date).getTime()) /
+                        ((bar.DT as Date).getTime() - (barPrev.DT as Date).getTime());
```
Comment (maryia-deriv) [inline src/store/ChartAdapterStore.ts:534]: just fixing 'Forbidden non-null assertion' TS warning

**Review by kate-deriv — APPROVED** [review]

**Review by ahmadtaimoor-deriv — APPROVED** [review]

**Review by kate-deriv — APPROVED** [review]

**Review by vinu-deriv — APPROVED** [review]

**Review by akmal-deriv — APPROVED** [review]

**Review by balakrishna-deriv — APPROVED** [review]

**Pull request merged**